exporters/otlp/otlplog/otlploggrpc/internal/observ/instrumentation.go

```diff
@@ -0,0 +1,269 @@
+const (
+	// ScopeName is the unique name of the meter used for instrumentation.
+	ScopeName = "go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploggrpc/internal/observ"
+
+	// Version is the current version of this instrumentation.
+	//
+	// This matches the version of the exporter.
+	Version = internal.Version
+)
+
+var (
+	attrsPool = &sync.Pool{
+		New: func() any {
+			// "component.name" + "component.type" + "error.type" + "server.address" + "server.port"
+			const n = 1 + 1 + 1 + 1 + 1
+			s := make([]attribute.KeyValue, 0, n)
+			return &s
+		},
+	}
+	addOpPool = &sync.Pool{
+		New: func() any {
+			const n = 1 // WithAttributeSet
+			o := make([]metric.AddOption, 0, n)
+			return &o
+		},
+	}
+	recordOptPool = &sync.Pool{
+		New: func() any {
+			const n = 1 + 1 // WithAttributeSet + "rpc.grpc.status_code"
+			o := make([]metric.RecordOption, 0, n)
+			return &o
+		},
+	}
+)
+
+func get[T any](p *sync.Pool) *[]T { return p.Get().(*[]T) }
+func put[T any](p *sync.Pool, s *[]T) {
+	*s = (*s)[:0]
+	p.Put(s)
+}
+
+// GetComponentName returns the constant name for the exporter with the
+// provided id.
+func GetComponentName(id int64) string {
+	return fmt.Sprintf("%s/%d", otelconv.ComponentTypeOtlpGRPCLogExporter, id)
+}
+
+// Instrumentation is experimental instrumentation for the exporter.
+type Instrumentation struct {
+	logInflightMetric         metric.Int64UpDownCounter
+	logExportedMetric         metric.Int64Counter
+	logExportedDurationMetric metric.Float64Histogram
+	presetAttrs               []attribute.KeyValue
+	setOpt                    metric.MeasurementOption
+}
+
+// NewInstrumentation returns instrumentation for otlplog grpc exporter.
+func NewInstrumentation(id int64, target string) (*Instrumentation, error) {
+	if !x.Observability.Enabled() {
+		return nil, nil
+	}
+
+	i := &Instrumentation{}
+
+	mp := otel.GetMeterProvider()
+	m := mp.Meter(
+		ScopeName,
+		metric.WithInstrumentationVersion(Version),
+		metric.WithSchemaURL(semconv.SchemaURL),
+	)
+
+	var err error
+
+	logInflightMetric, e := otelconv.NewSDKExporterLogInflight(m)
+	if e != nil {
+		e = fmt.Errorf("failed to create log inflight metric: %w", e)
+		err = errors.Join(err, e)
+	}
+	i.logInflightMetric = logInflightMetric.Inst()
+
+	logExportedMetric, e := otelconv.NewSDKExporterLogExported(m)
+	if e != nil {
+		e = fmt.Errorf("failed to create log exported metric: %w", e)
+		err = errors.Join(err, e)
+	}
+	i.logExportedMetric = logExportedMetric.Inst()
+
+	logOpDurationMetric, e := otelconv.NewSDKExporterOperationDuration(m)
+	if e != nil {
+		e = fmt.Errorf("failed to create log operation duration metric: %w", e)
+		err = errors.Join(err, e)
+	}
+	i.logExportedDurationMetric = logOpDurationMetric.Inst()
+	if err != nil {
+		return nil, err
+	}
+
+	i.presetAttrs = []attribute.KeyValue{
+		semconv.OTelComponentName(GetComponentName(id)),
+		semconv.OTelComponentTypeOtlpGRPCLogExporter,
+	}
+	i.presetAttrs = append(i.presetAttrs, ServerAddrAttrs(target)...)
```

Comment:
nit: this is likely going to over-allocate capacity for this slice. Merging the exact slice construction into a helper function will help avoid this by avoiding the `append`.